rpc/v9/dedup_cache.go

```diff
@@ -0,0 +1,113 @@
+package rpcv9
+
+import "sync"
+
+// Cache is expected to at most hold 3 blocks at a time.
+// latest block, pre-latest block, and the pre-confirmed block.
+const BlockCacheCapacity = 3
+
+// SubscriptionDeduplicationCache is a minimal cache for subscription notifications deduplication.
+// Stores notifications by block number and key, with automatic eviction.
+//
+// The cache is designed to hold at most 3 blocks:
+// - up to 2 non-committed blocks (pre-latest and pre-confirmed)
+// - and 1 committed block.
+// The finalised block will always have the smallest block number.
+// When the cache reaches capacity, it evicts the slot with the lowest block number, which always
+// corresponds to the oldest finalised block.
+type SubscriptionDeduplicationCache[K comparable, V comparable] struct {
+	mu    sync.RWMutex
+	slots [BlockCacheCapacity]struct {
+		blockNum uint64
+		data     map[K]V
+	}
+}
+
+// NewSubscriptionDeduplicationCache creates a new cache.
+func NewSubscriptionDeduplicationCache[
+	K comparable,
+	V comparable,
+]() *SubscriptionDeduplicationCache[K, V] {
+	cache := &SubscriptionDeduplicationCache[K, V]{}
+
+	for i := range cache.slots {
+		cache.slots[i].data = make(map[K]V)
+	}
+	return cache
+}
+
+// Put stores a notification for the given block number and key.
+// Evicts the lowest block number when cache is full.
+func (c *SubscriptionDeduplicationCache[K, V]) Put(blockNum uint64, key K, value V) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// If block exists, update in-place
+	if idx := c.findSlotIndex(blockNum); idx >= 0 {
+		c.slots[idx].data[key] = value
+		return
+	}
+
+	slotIdx := c.findSlotToUse()
+
+	clear(c.slots[slotIdx].data)
+	c.slots[slotIdx].blockNum = blockNum
+	c.slots[slotIdx].data[key] = value
+}
+
+// ShouldSend returns true if the notification should be sent.
+// Returns false if already cached for this block and key.
+func (c *SubscriptionDeduplicationCache[K, V]) ShouldSend(blockNum uint64, key K, value V) bool {
```

Comment:
This should receive the key and value by reference, given the use we are giving it to them. Small nitpick, since keys are `SentEvent` (5 words) and `felt.Hash` (4 words)